CMakeLists.txt

```diff
@@ -49,41 +51,43 @@ foreach (target cppzmq cppzmq-static)
                                                  $<INSTALL_INTERFACE:include>)
 endforeach()
 
-target_link_libraries(cppzmq INTERFACE libzmq)
-target_link_libraries(cppzmq-static INTERFACE libzmq-static)
-
-include(GNUInstallDirs)
-include(CMakePackageConfigHelpers)
-
-install(TARGETS cppzmq cppzmq-static
-        EXPORT ${PROJECT_NAME}-targets)
-
-install(FILES ${CPPZMQ_HEADERS}
-        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
-# GNUInstallDirs "DATADIR" wrong here; CMake search path wants "share".
-set(CPPZMQ_CMAKECONFIG_INSTALL_DIR "share/cmake/${PROJECT_NAME}" CACHE STRING "install path for cppzmqConfig.cmake")
-
-configure_file(libzmq-pkg-config/FindZeroMQ.cmake
-               libzmq-pkg-config/FindZeroMQ.cmake
-               COPYONLY)
-
-export(EXPORT ${PROJECT_NAME}-targets
-     FILE "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Targets.cmake")
-configure_package_config_file(${PROJECT_NAME}Config.cmake.in
-                              "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake"
-                              INSTALL_DESTINATION ${CPPZMQ_CMAKECONFIG_INSTALL_DIR})
-write_basic_package_version_file(${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake
-                                 VERSION ${CPPZMQ_VERSION}
-                                 COMPATIBILITY AnyNewerVersion)
-install(EXPORT ${PROJECT_NAME}-targets
-        FILE ${PROJECT_NAME}Targets.cmake
-        DESTINATION ${CPPZMQ_CMAKECONFIG_INSTALL_DIR})
-install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake
-              ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake
-              DESTINATION ${CPPZMQ_CMAKECONFIG_INSTALL_DIR})
-install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/libzmq-pkg-config/FindZeroMQ.cmake
-              DESTINATION ${CPPZMQ_CMAKECONFIG_INSTALL_DIR}/libzmq-pkg-config)
+target_link_libraries(cppzmq INTERFACE ${ZeroMQ_LIBRARY} )
+target_link_libraries(cppzmq-static INTERFACE ${ZeroMQ_STATIC_LIBRARY})
```

Comment:
I corrected what you mentioned in the new pull request. 

About the link part: the issue is that linking the target is typically the correct thing to do but you may want to be able to revert to the old style cmake variables. In fact I had that if/else but I forgot to commit it, sorry.
The deeper problem that requires that change is that you perform a search for ZeroMQ first which could trigger a FindModule in the user system which would set those two variables and set ZeroMQ to found but will not set the target appropriately. This is what happened to me on Windows installing ZeroMQ with vcpkg, as far as I remember and would happen anyway if one has his own FindModule in the search path. This way you should able to cope with that case as well

Reply:
@triskeldeian has a point here. I struggled with the same issue on Windows. I ended up writing my own `FindZeroMQ.cmake` and **intentionally** calling it _before_ the `find_package(cppzmq)`. Basically, I'm taking advantage of the mentioned ordering issue to thwart the problematic side-effects of the cppzmq config. I'm not sure this is quite the way I'd handle things myself, but the issue at the heart of this is real.

Reply:
Part of the problem is the non-standard way FindZeroMQ.cmake probes for the shared vs. static library, since they're separate-but-equal which means neither of them can really be considered "required". I ended up dropping the static lib alternative, in my own find module, since it isn't needed for any of my builds and getting rid of it makes everything _so_ much simpler. Like, I can use `FindPackageHandleStandardArgs` since `ZeroMQ_LIBRARY` is now one of the `REQUIRED_VARS`.

(Not using FPHSA in `FindZeroMQ.cmake` is the reason the `REQUIRED` argument has no effect in the `find_package()` call, BTW... just to address this source comment:)
https://github.com/zeromq/cppzmq/blob/18db4568f9ff1568193d3a496d87bed4bc9ec59e/CMakeLists.txt#L16-L22

Reply:
But, coming back around to this issue, the real problem is a quirk of how CMake's `target_link_libraries()` works. Since it's a heavily-overloaded call that's had a lot of different jobs over time, it's probably the least strict of all the CMake commands.

When a value is passed to it that isn't a target (say in a case of unconditional linking, when the target may or may not exist), it's just assumed to be an argument for the linker. CMake processes whatever it's given as a string arg that it will now happily stuff onto the linker command line when targets get linked to this one.

And if there's no target for it, then we can already be pretty certain that `libzmq` is _not_ a valid input for the linker, so having it there's gonna break some stuff.

Reply:
That's one of the reasons for namespacing `IMPORTED` targets, in fact. Because, while CMake can't be _sure_ that `libzmq` **isn't** a linker argument, it **CAN** be sure that e.g. `ZeroMQ::ZeroMQ` _must_ be a target, and otherwise it's not valid. (Same for `ZeroMQ::libzmq`, or any other permutation you like.) As the [`cmake-developer(7)`](https://cmake.org/cmake/help/v3.19/manual/cmake-developer.7.html) doc says:

> When providing imported targets, these should be namespaced (hence the `Foo::` prefix); CMake will recognize that values passed to `target_link_libraries` that contain `::` in their name are supposed to be imported targets (rather than just library names), and will produce appropriate diagnostic messages if that target does not exist (see policy `CMP0028`).